File: chatfield-py/chatfield/builder.py
"""Builder API for creating Chatfield interviews."""

import copy
from typing import Optional, List, Dict, Any, TYPE_CHECKING

# if TYPE_CHECKING:
from .interview import Interview


class TraitBuilder:
    """Builder for trait.possible() pattern."""
    
    def __init__(self, parent: 'RoleBuilder', role: str):
        self.parent = parent
        self.role = role
    
    def __call__(self, trait: str):
        """Add a regular trait."""
        self.parent._add_trait(self.role, trait)
        return self.parent
    
    def possible(self, name: str, trigger: str = ""):
        """Add a possible trait with optional trigger guidance."""
        self.parent._add_possible_trait(self.role, name, trigger)
        return self.parent


class RoleBuilder:
    """Builder for alice/bob role configuration."""
    
    def __init__(self, parent: 'ChatfieldBuilder', role: str):
        self.parent = parent
        self.role = role
        self._ensure_role()
        # Set context for subsequent calls
        self.parent._current_role = role
        self.trait = TraitBuilder(self, role)
    
    def _ensure_role(self):
        """Ensure role exists in chatfield structure."""
        if self.role not in self.parent._chatfield['roles']:
            self.parent._chatfield['roles'][self.role] = {
                'type': None,
                'traits': [],
                'possible_traits': {}
            }
    
    def type(self, role_type: str):
        """Set the role type."""
        self.parent._chatfield['roles'][self.role]['type'] = role_type
        return self
    
    def _add_trait(self, role: str, trait: str):
        """Add a regular trait."""
        if trait not in self.parent._chatfield['roles'][role]['traits']:
            self.parent._chatfield['roles'][role]['traits'].append(trait)
    
    def _add_possible_trait(self, role: str, name: str, trigger: str):
        """Add a possible trait."""
        self.parent._chatfield['roles'][role]['possible_traits'][name] = {
            'active': False,
            'desc': trigger
        }
    
    def field(self, name: str):
        """Start defining a new field."""
        return self.parent.field(name)
    
    def alice(self):
        """Switch to alice configuration."""
        return self.parent.alice()
    
    def bob(self):
        """Switch to bob configuration."""
        return self.parent.bob()
    
    def build(self):
        """Build the final Interview."""
        return self.parent.build()


class CastBuilder:
    """Builder for cast decorators with sub-attributes."""
    
    def __init__(self, parent: 'FieldBuilder', base_name: str, primitive_type: type, base_prompt: str):
        self.parent = parent
        self.base_name = base_name
        self.primitive_type = primitive_type
        self.base_prompt = base_prompt
    
    def __call__(self, prompt: Optional[str] = None):
        """Apply the base cast."""
        cast_info = {
            'type': self.primitive_type.__name__,
            'prompt': prompt or self.base_prompt
        }
        self.parent._chatfield_field['casts'][self.base_name] = cast_info
        return self.parent
    
    def __getattr__(self, name: str):
        """Handle sub-attributes like as_lang.fr."""
        if name.startswith('_'):
            raise AttributeError(f"No attribute {name}")
        
        compound_name = f'{self.base_name}_{name}'
        compound_prompt = self.base_prompt.format(name=name) if '{name}' in self.base_prompt else f'{self.base_prompt} for {name}'
        
        def sub_cast(prompt: Optional[str] = None):
            cast_info = {
                'type': self.primitive_type.__name__,
                'prompt': prompt or compound_prompt
            }
            self.parent._chatfield_field['casts'][compound_name] = cast_info
            return self.parent
        
        return sub_cast


class ChoiceBuilder:
    """Builder for choice cardinality decorators."""
    
    def __init__(self, parent: 'FieldBuilder', base_name: str, null: bool, multi: bool):
        self.parent = parent
        self.base_name = base_name
        self.null = null
        self.multi = multi
    
    def __getattr__(self, name: str):
        """Handle sub-attributes like as_one.parity."""
        if name.startswith('_'):
            raise AttributeError(f"No attribute {name}")
        
        compound_name = f'{self.base_name}_{name}'
        
        def choice_cast(*choices):
            cast_info = {
                'type': 'choice',
                'prompt': f"Choose for {name}",
                'choices': list(choices),
                'null': self.null,
                'multi': self.multi
            }
            self.parent._chatfield_field['casts'][compound_name] = cast_info
            return self.parent
        
        return choice_cast


class FieldBuilder:
    """Builder for individual fields."""
    
    def __init__(self, parent: 'ChatfieldBuilder', name: str):
        self.parent = parent
        self.name = name
        self._chatfield_field = {
            'desc': name,  # Default to field name
            'specs': {
                'must': [],
                'reject': [],
                'hint': [],
                'confidential': False,
                'conclude': False
            },
            'casts': {},
            'value': None
        }
        # Add to parent's fields
        self.parent._chatfield['fields'][name] = self._chatfield_field
        self.parent._current_field = name
        
        # Initialize cast builders
        self.as_int = CastBuilder(self, 'as_int', int, 'Parse as integer')
        self.as_float = CastBuilder(self, 'as_float', float, 'Parse as floating point number')
        self.as_bool = CastBuilder(self, 'as_bool', bool, 'Parse as boolean')
        self.as_str = CastBuilder(self, 'as_str', str, 'Format as string')
        self.as_percent = CastBuilder(self, 'as_percent', float, 'Parse as percentage (0.0 to 1.0)')
        self.as_list = CastBuilder(self, 'as_list', list, 'Parse as list/array')
        self.as_set = CastBuilder(self, 'as_set', set, 'Parse as unique set')
        self.as_dict = CastBuilder(self, 'as_dict', dict, 'Parse as key-value dictionary')
        self.as_obj = self.as_dict  # Alias
        self.as_lang = CastBuilder(self, 'as_lang', str, 'Translate to {name}')
        
        # Choice builders
        self.as_one = ChoiceBuilder(self, 'as_one', null=False, multi=False)
        self.as_maybe = ChoiceBuilder(self, 'as_maybe', null=True, multi=False)
        self.as_multi = ChoiceBuilder(self, 'as_multi', null=False, multi=True)
        self.as_any = ChoiceBuilder(self, 'as_any', null=True, multi=True)
    
    def desc(self, description: str):
        """Set field description."""
        self._chatfield_field['desc'] = description
        return self
    
    def must(self, rule: str):
        """Add a validation requirement."""
        self._chatfield_field['specs']['must'].append(rule)
        return self
    
    def reject(self, rule: str):
        """Add a rejection rule."""
        self._chatfield_field['specs']['reject'].append(rule)
        return self
    
    def hint(self, tooltip: str):
        """Add helpful context."""
        if not self._chatfield_field['specs'].get('hint'):
            self._chatfield_field['specs']['hint'] = []
        self._chatfield_field['specs']['hint'].append(tooltip)
        return self
    
    def confidential(self):
        """Mark field as confidential (tracked silently)."""
        self._chatfield_field['specs']['confidential'] = True
        return self
    
    def conclude(self):
        """Mark field for evaluation only after conversation ends (automatically confidential)."""
        self._chatfield_field['specs']['conclude'] = True
        self._chatfield_field['specs']['confidential'] = True  # Implied
        return self
    
    
    def field(self, name: str):
        """Start defining a new field."""
        return self.parent.field(name)
    
    def alice(self):
        """Switch to alice configuration."""
        return self.parent.alice()
    
    def bob(self):
        """Switch to bob configuration."""
        return self.parent.bob()
    
    def build(self):
        """Build the final Interview."""
        return self.parent.build()


class ChatfieldBuilder:
    """Main builder for creating Chatfield interviews."""
    
    def __init__(self):
        self._chatfield = {
            'type': '',
            'desc': '',
            'roles': {},
            'fields': {}
        }
        self._current_field: Optional[str] = None
        self._current_role: Optional[str] = None
    
    def type(self, name: str):
        """Set the interview type name."""
        self._chatfield['type'] = name
        return self
    
    def desc(self, description: str):
        """Set the interview description."""
        self._chatfield['desc'] = description
        return self
    
    def alice(self):
        """Configure alice (interviewer) role."""
        return RoleBuilder(self, 'alice')
    
    def bob(self):
        """Configure bob (interviewee) role."""
        return RoleBuilder(self, 'bob')
    
    def field(self, name: str):
        """Define a new field."""
        return FieldBuilder(self, name)
    
    def build(self):
        """Build the final Interview object."""
        # Create Interview instance with the built structure and override its _chatfield.
        interview = Interview()
        interview._chatfield = copy.deepcopy(self._chatfield)
        
        # Ensure roles are properly initialized with defaults
        if 'alice' not in interview._chatfield['roles']:
            interview._chatfield['roles']['alice'] = {
                'type': 'Agent',
                'traits': [],
                'possible_traits': {}
            }
        if 'bob' not in interview._chatfield['roles']:
            interview._chatfield['roles']['bob'] = {
                'type': 'User', 
                'traits': [],
                'possible_traits': {}
            }
        
        # Ensure possible_traits dict exists for each role
        for role in interview._chatfield['roles'].values():
            if 'possible_traits' not in role:
                role['possible_traits'] = {}
        
        return interview


def chatfield():
    """Create a new Chatfield builder."""
    return ChatfieldBuilder()


# Preset builders for common patterns
def patient_gatherer():
    """Create a patient, thorough gatherer."""
    return (chatfield()
        .alice()
            .trait("patient and thorough")
            .trait("asks follow-up questions when answers seem incomplete")
            .trait("provides helpful examples when users seem confused"))


def quick_gatherer():
    """Create a quick, efficient gatherer."""
    return (chatfield()
        .alice()
            .trait("concise and efficient")
            .trait("accepts brief answers when they meet requirements")
            .trait("moves quickly through fields"))


def expert_gatherer():
    """Create an expert consultation gatherer."""
    return (chatfield()
        .alice()
            .trait("assumes domain expertise")
            .trait("asks detailed technical questions")
            .trait("expects comprehensive, specific answers"))

File: chatfield-js/src/builder.ts
/**
 * Type-safe Builder API for creating Chatfield interviews.
 * Mirrors Python's builder.py with full TypeScript type tracking.
 */

import { Interview } from './interview'
import type {
  FieldMeta,
  FieldSpecs,
  InterviewMeta,
  RoleMeta,
  TraitBuilder,
  CastBuilder,
  ChoiceBuilder,
  CastInfo
} from './builder-types'

/**
 * Creates a callable TraitBuilder
 */
function createTraitBuilder(parent: RoleBuilder, role: string): TraitBuilder {
  const addTrait = (trait: string) => {
    parent._addTrait(role, trait)
    return parent
  }
  
  addTrait.possible = (name: string, trigger: string = "") => {
    parent._addPossibleTrait(role, name, trigger)
    return parent
  }
  
  return addTrait
}

/**
 * Creates a callable CastBuilder with function-based API
 */
function createCastBuilder<TParent extends FieldBuilder<any, any>>(
  parent: TParent,
  baseName: string,
  primitiveType: any,
  basePrompt: string,
  requiresSubName: boolean = false
): CastBuilder<TParent> {
  // Function with overloads for different parameter combinations
  const castFunction = function(...args: any[]): TParent {
    let subName: string | undefined
    let customPrompt: string | undefined
    
    if (requiresSubName) {
      // Mandatory sub-name (like as_lang)
      // as_lang("fr") or as_lang("fr", "Translate to French")
      if (args.length === 0) {
        throw new Error(`${baseName} requires a sub-name parameter`)
      }
      subName = String(args[0])
      customPrompt = args[1] ? String(args[1]) : undefined
    } else {
      // Optional sub-name (like as_int)
      if (args.length === 0) {
        // as_int() - default behavior
      } else if (args.length === 1) {
        // as_int("custom prompt") - override prompt
        customPrompt = String(args[0])
      } else if (args.length >= 2) {
        // as_int("severity", "Range 0-10") - sub-name + prompt
        subName = String(args[0])
        customPrompt = String(args[1])
      }
    }
    
    // Build the cast name
    const castName = subName ? `${baseName}_${subName}` : baseName
    
    // Build the prompt
    let finalPrompt: string
    if (customPrompt) {
      finalPrompt = customPrompt
    } else if (subName && basePrompt.includes('{name}')) {
      finalPrompt = basePrompt.replace('{name}', subName)
    } else if (subName) {
      finalPrompt = `${basePrompt} for ${subName}`
    } else {
      finalPrompt = basePrompt
    }
    
    // Store the cast
    const castInfo: CastInfo = {
      type: typeof primitiveType === 'function' ? primitiveType.name.toLowerCase() : String(primitiveType),
      prompt: finalPrompt
    }
    parent._chatfieldField.casts[castName] = castInfo
    return parent
  }
  
  return castFunction as CastBuilder<TParent>
}

/**
 * Creates a callable ChoiceBuilder with function-based API
 */
function createChoiceBuilder<TParent extends FieldBuilder<any, any>>(
  parent: TParent,
  baseName: string,
  nullAllowed: boolean,
  multi: boolean
): ChoiceBuilder<TParent> {
  const choiceFunction = function(...args: any[]): TParent {
    let subName: string | undefined
    let choices: string[]
    
    // Check if first arg is a sub-name (non-choice string followed by choices)
    if (args.length >= 2 && typeof args[0] === 'string') {
      // Could be as_one('selection', 'red', 'green', 'blue')
      // We'll treat it as sub-name if followed by more strings
      const possibleChoices = args.slice(1)
      if (possibleChoices.every(arg => typeof arg === 'string')) {
        subName = args[0]
        choices = possibleChoices
      } else {
        // All args are choices
        choices = args
      }
    } else {
      // Direct choices: as_one('red', 'green', 'blue')
      choices = args
    }
    
    // Build the cast name
    const castName = subName ? `${baseName}_${subName}` : baseName
    
    // Build the prompt
    const promptBase = `Choose ${multi ? 'one or more' : 'one'}`
    const prompt = subName ? `${promptBase} for ${subName}` : promptBase
    
    // Store the cast
    const castInfo: CastInfo = {
      type: 'choice',
      prompt: prompt,
      choices: choices,
      null: nullAllowed,
      multi: multi
    }
    parent._chatfieldField.casts[castName] = castInfo
    return parent
  }
  
  return choiceFunction as ChoiceBuilder<TParent>
}

/**
 * Builder for alice/bob role configuration.
 * Mirrors Python's RoleBuilder
 */
export class RoleBuilder<Fields extends string = never> {
  parent: ChatfieldBuilder<Fields>
  role: string
  trait: TraitBuilder

  constructor(parent: ChatfieldBuilder<Fields>, role: string) {
    this.parent = parent
    this.role = role
    this._ensureRole()
    // Set context for subsequent calls
    this.parent._currentRole = role
    this.trait = createTraitBuilder(this, role)
  }

  private _ensureRole(): void {
    // Ensure role exists in chatfield structure
    if (!(this.role in this.parent._chatfield.roles)) {
      this.parent._chatfield.roles[this.role] = {
        type: null,
        traits: [],
        possible_traits: {}
      }
    }
  }

  type(roleType: string): RoleBuilder<Fields> {
    // Set the role type
    const roleData = this.parent._chatfield.roles[this.role]
    if (roleData) {
      roleData.type = roleType
    }
    return this
  }

  _addTrait(role: string, trait: string): void {
    // Add a regular trait
    const roleData = this.parent._chatfield.roles[role]
    if (roleData && !roleData.traits.includes(trait)) {
      roleData.traits.push(trait)
    }
  }

  _addPossibleTrait(role: string, name: string, trigger: string): void {
    // Add a possible trait
    const roleData = this.parent._chatfield.roles[role]
    if (roleData && roleData.possible_traits) {
      roleData.possible_traits[name] = {
        active: false,
        desc: trigger
      }
    }
  }

  field<Name extends string>(name: Name): FieldBuilder<Fields | Name, Name> {
    // Start defining a new field
    return this.parent.field(name)
  }

  alice(): RoleBuilder<Fields> {
    // Switch to alice configuration
    return this.parent.alice()
  }

  bob(): RoleBuilder<Fields> {
    // Switch to bob configuration
    return this.parent.bob()
  }

  build(): TypedInterview<Fields> {
    // Build the final Interview
    return this.parent.build()
  }
}

/**
 * Builder for field configuration with all decorators.
 * Tracks the field name in the type system.
 */
export class FieldBuilder<Fields extends string, CurrentField extends string> {
  parent: ChatfieldBuilder<Fields>
  fieldName: CurrentField
  _chatfieldField: FieldMeta

  // Cast builders
  as_int: CastBuilder<this>
  as_float: CastBuilder<this>
  as_bool: CastBuilder<this>
  as_percent: CastBuilder<this>
  as_lang: CastBuilder<this>
  as_str: CastBuilder<this>
  as_one: ChoiceBuilder<this>
  as_maybe: ChoiceBuilder<this>
  as_multi: ChoiceBuilder<this>
  as_any: ChoiceBuilder<this>

  constructor(parent: ChatfieldBuilder<Fields>, fieldName: CurrentField) {
    this.parent = parent
    this.fieldName = fieldName
    
    // Ensure field exists
    const fieldKey = fieldName as unknown as Fields
    if (!(fieldKey in this.parent._chatfield.fields)) {
      (this.parent._chatfield.fields as any)[fieldKey] = {
        desc: '',
        specs: {
          must: [],
          reject: [],
          hint: []
        },
        casts: {},
        value: null
      }
    }
    
    this._chatfieldField = (this.parent._chatfield.fields as any)[fieldKey]
    
    // Set as current field
    this.parent._currentField = fieldName as string
    
    // Initialize cast builders
    // Optional sub-name casts (can be called with 0, 1, or 2 args)
    this.as_int = createCastBuilder(this, 'as_int', Number, 'parse as integer', false)
    this.as_float = createCastBuilder(this, 'as_float', Number, 'parse as float', false)
    this.as_bool = createCastBuilder(this, 'as_bool', Boolean, 'parse as boolean', false)
    this.as_percent = createCastBuilder(this, 'as_percent', Number, 'parse as percentage (0-100)', false)
    // Mandatory sub-name cast (requires at least 1 arg)
    this.as_lang = createCastBuilder(this, 'as_lang', String, 'translate to {name}', true)
    // Add as_str for custom transformations
    this.as_str = createCastBuilder(this, 'as_str', String, 'format as {name}', false)
    
    // Initialize choice builders
    this.as_one = createChoiceBuilder(this, 'as_one', false, false)
    this.as_maybe = createChoiceBuilder(this, 'as_maybe', true, false)
    this.as_multi = createChoiceBuilder(this, 'as_multi', false, true)
    this.as_any = createChoiceBuilder(this, 'as_any', true, true)
  }

  desc(description: string): this {
    // Set field description
    this._chatfieldField.desc = description
    return this
  }

  must(rule: string): this {
    // Add a validation requirement
    if (!this._chatfieldField.specs.must) {
      this._chatfieldField.specs.must = []
    }
    this._chatfieldField.specs.must.push(rule)
    return this
  }

  reject(rule: string): this {
    // Add a rejection rule
    if (!this._chatfieldField.specs.reject) {
      this._chatfieldField.specs.reject = []
    }
    this._chatfieldField.specs.reject.push(rule)
    return this
  }

  hint(tooltip: string): this {
    // Add helpful context
    if (!this._chatfieldField.specs.hint) {
      this._chatfieldField.specs.hint = []
    }
    this._chatfieldField.specs.hint.push(tooltip)
    return this
  }

  confidential(): this {
    // Mark field as confidential (tracked silently)
    this._chatfieldField.specs.confidential = true
    return this
  }

  conclude(): this {
    // Mark field for evaluation only after conversation ends (automatically confidential)
    this._chatfieldField.specs.conclude = true
    this._chatfieldField.specs.confidential = true  // Implied
    return this
  }

  field<Name extends string>(name: Name): FieldBuilder<Fields | Name, Name> {
    // Start defining a new field
    return this.parent.field(name)
  }

  alice(): RoleBuilder<Fields> {
    // Switch to alice configuration
    return this.parent.alice()
  }

  bob(): RoleBuilder<Fields> {
    // Switch to bob configuration
    return this.parent.bob()
  }

  build(): TypedInterview<Fields> {
    // Build the final Interview
    return this.parent.build()
  }
}

/**
 * Main builder for creating Chatfield interviews with type tracking.
 * The Fields generic tracks all field names added during building.
 */
export class ChatfieldBuilder<Fields extends string = never> {
  _chatfield: InterviewMeta<Fields>
  _currentField: string | null
  _currentRole: string | null

  constructor() {
    this._chatfield = {
      type: '',
      desc: '',
      roles: {
        alice: {
          type: null,
          traits: [],
          possible_traits: {}
        },
        bob: {
          type: null,
          traits: [],
          possible_traits: {}
        }
      },
      fields: {} as Record<Fields, FieldMeta>
    }
    this._currentField = null
    this._currentRole = null
  }

  type(name: string): ChatfieldBuilder<Fields> {
    // Set the interview type name
    this._chatfield.type = name
    return this
  }

  desc(description: string): ChatfieldBuilder<Fields> {
    // Set the interview description
    this._chatfield.desc = description
    return this
  }

  alice(): RoleBuilder<Fields> {
    // Configure alice (interviewer) role
    return new RoleBuilder(this, 'alice')
  }

  bob(): RoleBuilder<Fields> {
    // Configure bob (interviewee) role
    return new RoleBuilder(this, 'bob')
  }

  field<Name extends string>(name: Name): FieldBuilder<Fields | Name, Name> {
    // Start defining a new field, tracking its name in the type system
    return new FieldBuilder(this as any as ChatfieldBuilder<Fields | Name>, name)
  }

  build(): TypedInterview<Fields> {
    // Build the final Interview with known field names
    const interview = new Interview(
      this._chatfield.type,
      this._chatfield.desc,
      this._chatfield.roles,
      this._chatfield.fields
    )
    return interview as TypedInterview<Fields>
  }
}

/**
 * Type-safe Interview with known field names
 */
export type TypedInterview<Fields extends string> = Interview & {
  [K in Fields]: string  // Each field is accessible as a string property
}

/**
 * Main entry point - creates a new type-safe builder
 * @returns A new ChatfieldBuilder instance
 */
export function chatfield<Fields extends string = never>(): ChatfieldBuilder<Fields> {
  return new ChatfieldBuilder<Fields>()
}

/**
 * Loose type version for dynamic field names (Python-like experience)
 */
export function chatfieldDynamic(): ChatfieldBuilder<string> {
  return new ChatfieldBuilder<string>()
}

/**
 * Preset builders for common patterns
 */

/**
 * Create a patient, thorough gatherer.
 * Mirrors Python's patient_gatherer()
 */
export function patientGatherer(): ChatfieldBuilder<string> {
  const builder = chatfield<string>()
    .alice()
  builder.trait('Takes time to understand')
  builder.trait('Asks clarifying questions when responses are unclear or incomplete')
  builder.trait('Guides the conversation with gentle suggestions if needed')
  return builder.parent
}

/**
 * Create a quick, efficient gatherer.
 * Mirrors Python's quick_gatherer()
 */
export function quickGatherer(): ChatfieldBuilder<string> {
  const builder = chatfield<string>()
    .alice()
  builder.trait('Gets to the point quickly')
  builder.trait('Focuses on essential information without extra conversation')
  builder.trait('Efficient and business-like')
  return builder.parent
}

/**
 * Create an expert consultation gatherer.
 * Mirrors Python's expert_gatherer()
 */
export function expertGatherer(): ChatfieldBuilder<string> {
  const builder = chatfield<string>()
    .alice()
  builder.trait('Acts as a domain expert')
  builder.trait('Provides context and education during the conversation')
  builder.trait('Can explain complex topics when relevant to the discussion')
  return builder.parent
}